Add KMS plugin registry

sakshamsharma opened the pull request:
Allows supporting KMS services as encryption providers using a plugin mechanism similar to admission plugins.

Simplifies #48574 

Progresses #48522

@deads2k PTAL

sakshamsharma committed:
Add KMS plugin registry

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go:
```diff
@@ -69,3 +71,14 @@ type Key struct {
+	// +optional
+	CacheSize int `json:"cachesize,omitempty"`
+	// configfile is the path to the configuration file for the named KMS provider.
+	ConfigFile string `json:"configfile"`
```
I'd like to see this structure follow the admission plugin configuration structure and allow either a path or a rawextension directly here.

deads2k replied:
> I'd like to see this structure follow the admission plugin configuration structure and allow either a path or a rawextension directly here.

This is ok as a followup

sakshamsharma commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/plugins.go:
```diff
@@ -0,0 +1,122 @@
+	if err != nil {
+		return nil, true, err
+	}
+	if !PluginEnabledFn(name, config1) {
```
I am not entirely sure whether this part is relevant to this process, and I think it can be removed without introducing any immediate issues.

deads2k replied:
> I am not entirely sure whether this part is relevant to this process, and I think it can be removed without introducing any immediate issues. Have copied this from admissions plugins code.

You should probably leave it for zero config options.

sakshamsharma replied:
> You should probably leave it for zero config options.

Does that mean I should delete this? I'm sorry, I didn't understand `zero config options`

deads2k commented:
I'd like to see this moved to be a peer of AdmissionConfiguration and gain proper validation functions.

There is code duplication that should be resolved, but as a building block, I think this keeps our types closer to where they need to be.  

The actual handling of the `EncryptionProviderConfigFilepath` needs to move into the storage layer apply methods too.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go:
```diff
@@ -69,3 +71,14 @@ type Key struct {
+	Name string `json:"name"`
+	// cacheSize is the maximum number of secrets which are cached in memory. The default value is 1000.
+	// +optional
+	CacheSize int `json:"cachesize,omitempty"`
```
shouldn't cachesize be part of the config?

sakshamsharma replied:
It is not specific to the plugin's configuration. It is a property of the envelope transformer. It will also be used by cloudprovidedkms based plugins.

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -38,6 +39,7 @@ const (
 	aesCBCTransformerPrefixV1    = "k8s:enc:aescbc:v1:"
 	aesGCMTransformerPrefixV1    = "k8s:enc:aesgcm:v1:"
 	secretboxTransformerPrefixV1 = "k8s:enc:secretbox:v1:"
+	envelopeTransformerPrefixV1  = "k8s:enc:envelope:v1:"
```
seems like this should match the config field name, both `envelope` or both `kms`

deads2k commented:
I think this moves us in the right direction in spite of the debt.  Fix the prefix and I can get behind this one.

sakshamsharma committed:
Add unit tests for KMS transformer initialization

sakshamsharma commented:
@lavalamp @cheftako Can you PTAL at this? The idea is to mirror the admission plugins code.
So no new parameter would be passed from kube-apiserver to the generic apiserver encryption configuration parsing.
We would register the cloud from the kube-apiserver. Other KMS providers like vault will register themselves to this plugin factory as well.

cheftako commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -144,6 +146,27 @@ func GetPrefixTransformers(config *ResourceConfig) ([]value.PrefixTransformer, e
+			}
+			defer f.Close()
+			envelopeService, pluginFound, err := KMSPluginRegistry.getPlugin(provider.KMS.Name, f)
+			if pluginFound == false {
```
Please reorder found vs err handling. Assuming that there is an actual error, reporting that should take precedence. Especially as the failure to find a plugin provider could be the result of the error.

sakshamsharma commented:
/release-note-none

cheftako commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/plugins.go:
```diff
@@ -0,0 +1,122 @@
+
+// All registered KMS options.
+type KMSPlugins struct {
+	lock     sync.Mutex
```
I would expect that registration is a rare event but that getting a plugin might not be. If so then this might benefit from a RWLock.

cheftako commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/plugins.go:
```diff
@@ -0,0 +1,122 @@
+	if found {
+		glog.Fatalf("KMS plugin %q was registered twice", name)
+	}
+	if ps.registry == nil {
```
This works but it feels wrong to me to access a map just prior to initializing. May be less work on later devs if you put the nil check first.

cheftako commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/plugins.go:
```diff
@@ -0,0 +1,122 @@
+// known but failed to initialize.  The config parameter specifies the io.Reader
+// handler of the configuration file for the cloud provider, or nil for no configuration.
+func (ps *KMSPlugins) getPlugin(name string, config io.Reader) (envelope.Service, bool, error) {
+	ps.lock.Lock()
```
Not sure it is a good idea to hold the lock for this entire method. As it stands there are multiple pieces of code for which you are holding the lock that you do not control and could even be RPC calls. Assuming that the lock is only securing access to the registry object, I would suggest you should create a smaller get method which just access registry with its lock and then call that from here.

cheftako reviewed (commented):
Minor tweaks

sakshamsharma committed:
Add cloudprovidedkms provider support

sakshamsharma commented:
@cheftako Addressed those issues. I also added support for cloudprovidedkms configuration. It looks almost the same as the KMS one. Another quick pass would be greatly appreciated, since the cloudprovidedkms was something you had recommended.

sakshamsharma commented:
/test pull-kubernetes-e2e-kops-aws

deads2k commented:
> @deads2k @liggitt if this is still acceptable for you, can we merge this in? The other PR can then get in too (basically the changes on the above mentioned branch on my fork) because the controversial part is covered in this one.

I'll merge this bit this it is directly related to the API machinery and gives a good (if slightly debt-y) plugin mechanism.  The other implementation pull seems to have slipped past sig-auth until a couple days ago.  There probably ought to be more time to comment there.

/lgtm

cheftako commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -144,6 +147,38 @@ func GetPrefixTransformers(config *ResourceConfig) ([]value.PrefixTransformer, e
+			if found == true {
+				return nil, fmt.Errorf("more than one provider specified in a single element, should split into different list elements")
+			}
+			f, err := os.Open(provider.KMS.ConfigFile)
```
It feels wrong that we are trying to open/read the file here. It would seem better to just pass this location down to the factory and let it do the read. Especially as in something like a hybrid case this could be a key into a config server or be a file+coordinate into a larger config file. We should need to know or care about such things. However if we insist on opening+reading the file then we end up having to care about these things.

sakshamsharma replied:
The admissions factory used this format, which is why I stuck with this. We can change this in a follow-up.
Once this PR is merged, I'll open a new one to change this and remove the reflect from your 2nd comment, along with some other cleanup.

deads2k replied:
> It feels wrong that we are trying to open/read the file here. It would seem better to just pass this location down to the factory and let it do the read. Especially as in something like a

To make this API congruent to the admission config, there will be a choice of separate file or embedded content depending the sensitivity (which we don't know since its pluggable) of the config in question.

sakshamsharma replied:
An approach which I've proposed before, and I've seen used in the cloudproviders package:
Have a field `config map[string]interface{}` on the `KMSConfig` struct. Send it to the plugin factory. The plugin factory would use the `mapstructure` library to parse it into a struct safely.

So this way you can support both `config:` and `configfile:` attributes on the `kms` configuration. Just putting this out here.

liggitt replied:
The reader interface lets you load a structured config from an external file or an inlined config. All our config files should be structured and versioned (see scheduler config, ABAC config, admission config, etc)

cheftako commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/plugins.go:
```diff
@@ -0,0 +1,132 @@
+
+// splitStream reads the stream bytes and constructs two copies of it.
+func splitStream(config io.Reader) (io.Reader, io.Reader, error) {
+	if config == nil || reflect.ValueOf(config).IsNil() {
```
What are we trying to protect ourselves against with this use of reflect? None of our other nil checks use it (other than the place we cut&pasted this code from).  When we opened the config reader a nil check was sufficient. This object has not been copied/valueOf so I believe we are pulling in something we are better of without for no good reason.

deads2k replied:
> What are we trying to protect ourselves against with this use of reflect? None of our other nil checks use it (other than the place we cut&pasted this code from). When we opened the config reader a nil check was sufficient. This object has not been copied/valueOf so I believe we are pulling in something we are better of without for no good reason

Nil versus nil.  Is it a nil interface or a non-nil interface with a nil value.  It may drop out when these plugin paths are collapsed.

cheftako reviewed (approved):
Would like to see us fix config handling in a later release. Might be good to file an issue on it.

sakshamsharma commented:
One of the comments was not in the proper format, and this package was removed from `hack/.lint_failures` by another PR recently.

deads2k commented:
/retest

cjwagner commented:
bump

cjwagner commented:
/test pull-kubernetes-unit
